db/commitlog/commitlog_replayer.cc

```diff
@@ -112,65 +112,46 @@ db::commitlog_replayer::impl::impl(seastar::sharded<replica::database>& db, seas
 {}
 
 future<> db::commitlog_replayer::impl::init() {
-    return _db.map_reduce([this](shard_rpm_map map) {
-        for (auto& p1 : map) {
-            for (auto& p2 : p1.second) {
-                auto& pp = _rpm[p1.first][p2.first];
-                pp = std::max(pp, p2.second);
-
-                auto i = _min_pos.find(p1.first);
-                if (i == _min_pos.end() || p2.second < i->second) {
-                    _min_pos[p1.first] = p2.second;
```

Comment:
From the commit message I understand why we don't need to gather results from other shards. However, the `map` part of this is **mutating** the structures. We don't need those mutations to happen?

I also have trouble understanding what the original code does, variable names like `p1`, `p2`, `pp`, and zero comments don't make it easy...

Reply:
In the previous version we got an identical maps of the shape `shard -> table_id -> truncation_RP` from each shard and in `map` part we constructed `_rpm` and `_min_pos` based on them. In the new version we do the same once we read the truncation positions.

Reply:
For the record, the reason the code was a map-reduce is that originally we also collected replay positions from sstables, to help determine the "min_pos" and potentially eliminate segments early. This of course must be collected across all shards, hence the now redundant map-reduce dance. 

This was removed because we could did not reliably retain/propagate these replay positions, esp. across compactions (even though they are only valid for a single boot-cycle). At some point, maybe we will be better (already are?), but as it stands today, we don't really eliminate much here.